pkg/infrastructure/clusterapi/clusterapi.go

```diff
@@ -0,0 +1,363 @@
+package clusterapi
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	utilkubeconfig "sigs.k8s.io/cluster-api/util/kubeconfig"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/cluster/metadata"
+	"github.com/openshift/installer/pkg/asset/ignition/bootstrap"
+	"github.com/openshift/installer/pkg/asset/ignition/machine"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	"github.com/openshift/installer/pkg/asset/kubeconfig"
+	"github.com/openshift/installer/pkg/asset/machines"
+	"github.com/openshift/installer/pkg/asset/manifests/capiutils"
+	capimanifests "github.com/openshift/installer/pkg/asset/manifests/clusterapi"
+	"github.com/openshift/installer/pkg/clusterapi"
+	"github.com/openshift/installer/pkg/infrastructure"
+	"github.com/openshift/installer/pkg/types"
+)
+
+// InfraProvider implements the infrastructure/provider interface
+// which is the interface between the rest of the installer and
+// the CAPI infrastructure provider.
+type InfraProvider struct {
+	infrastructure.Provider
+
+	capiProvider Provider
+}
+
+// InitializeProvider returns a CAPI provider implementation for a specific
+// cloud platform.
+func InitializeProvider(platform Provider) infrastructure.Provider {
+	return InfraProvider{capiProvider: platform}
+}
+
+// Provider is the base interface that cloud platforms
+// should implement for the CAPI infrastructure provider.
+type Provider interface {
+	// Name provides the name for the cloud platform.
+	Name() string
+}
+
+// PreProvider can be implemented by cloud platforms to perform functions before
+// CAPI provisioning has begun.
+type PreProvider interface {
+	// PreProvision is called before provisioning using CAPI controllers has begun
+	// and should be used to create dependencies needed for CAPI provisioning,
+	// such as IAM roles or policies.
+	PreProvision(ctx context.Context, in PreProvisionInput) error
+}
+
+// PreProvisionInput collects the args to be passed to the PreProvision call.
+type PreProvisionInput struct {
+	InfraID       string
+	InstallConfig *installconfig.InstallConfig
+}
+
+// IgnitionProvider generates the ignition secrets for the bootstrap machine
+// and handles any preconditions for ignition.
+type IgnitionProvider interface {
+	Ignition(ctx context.Context, in IgnitionInput) (client.Object, error)
```

Comment:
```suggestion
	Ignition(ctx context.Context, in IgnitionInput) ([]byte, error)
```

I'd expect this interface to give the Ignition data as input and return Ignition data as output, in the default case this should be a no-op. The secret should still be created outside the interface with the bytes returned.

This way, the ignition data can be modified, although we should highly discourage the use of this interface

Reply:
fine by me! Even less room for error this way.

Reply:
Done. Thanks!